Add testing guide for Jest and Mocha

cowboyd opened the pull request:
## Motivation
Our unit testing integration is one of the most powerful aspects of Effection, but right now there is not much documentation on the concepts of how to make it work.

## Approach
This adds a general testing strategy for how to do testing with effection, followed by a mini-tutorial on using it with Jest and Mocha since we a) have those already and b) they are so close together that we can speak both audiences of users together.

## Preview

https://deploy-preview-616--effection.netlify.app/effection/docs/guides/testing

cowboyd committed:
Add testing guide for Jest and Mocha

Our unit testing integration is one of the most powerful aspects of
Effection, but right now there is not much documentation on the
concepts of how to make it work.

This adds a general testing strategy for how to do testing with
effection, followed by a mini-tutorial on using it with Jest and
Mocha since we a) have those already and b) they are so close together
that we can speak both audiences of users together.

dagda1 commented on website/docs/guides/testing.md:
```diff
@@ -3,70 +3,486 @@ id: testing
+While there are certainly different ways of expressing this syntactically, it
+turns out that as a fundamental testing pattern, it is nearly universal. Notice
+that, like tests, the cleanup of an effection operation is intrinsic. We can
+leverage this fact to associate an effection task, or scope, with each test.
```
You might want to link to a doc about the nomenclature (if there is one).  The reader might not know what a task, scope or resource is.

cowboyd replied:
@dagda1 I made changes to keep all of the discussion referencing the concrete concept of tasks, and halting tasks instead of "scopes" or "disposing of scope"

dagda1 commented on website/docs/guides/testing.md:
```diff
@@ -3,70 +3,486 @@ id: testing
+concise, but also more flexible. To do this, we use the fact that each test
+case gets its own scope that is automatically destroyed for us after it runs.
+
+So if we re-cast our server as a [resource][] that is running in our test scope,
```
no link in `resource[]`

cowboyd replied:
It's at the bottom of the document. Turns out Markdown uses the text as the link if one is not provided.

dagda1 commented on website/docs/guides/testing.md:
```diff
@@ -3,70 +3,486 @@ id: testing
+import { beforeEach, it } from '@effection/jest';
+
+describe("a server", () => {
+  beforeEach(function*() {
```
I like this example

dagda1 reviewed (changes requested):
Looks good, a couple of small comments.

cowboyd committed:
👍Convert references to "scope" to use "task"

In Effection, the actual thing is a task that manages scope, so in
order to be unambiguous, let's refer to task when we're talking about
the actual thing which is created, and halted.

In a couple of cases, I've called the task associated with the suite
the "suite-scoped" task, and the task with the lifetime of the `it`
block the "test-scoped" task. In both cases, the thing we're talking
about is a task.

cowboyd committed:
👍Further clarification of scope

dagda1 commented on website/docs/guides/testing.md:
```diff
@@ -365,12 +367,13 @@ describe("a server", () => {
-the same token, the _test_ scope is disposed of after every single test is
-finished. Any tasks spawned within it will be halted immediately after the test
-is finished. For example:
+As hinted at above, there are two separate tasks scopes at play in
```
> two separate tasks scopes at play

I'm not sure what this means.  Should it be either `tasks` or `scopes`

dagda1 commented on website/docs/guides/testing.md:
```diff
@@ -365,12 +367,13 @@ describe("a server", () => {
-is finished. For example:
+As hinted at above, there are two separate tasks scopes at play in
+your tests: _suite_ scope, and _test_ scope. The lifetime of the suite
+scope task is the entire test run.  Any task spawned within it can
```
`scope` and `task` together

cowboyd replied:
Oops, that didn't read well did it. Let me have another go.

cowboyd committed:
👍 make sure to call tasks tasks

Only use "test-scoped" and "suite-scoped" as adjectives for task.

dagda1 reviewed (approved):
This reads much better now.  Good work!

taras reviewed (approved):
Agree with @dagda1, it reads very well now
